### ISSUE TITLE
VSCode syntax highlighting not working for my project

### ISSUE THREAD

**Issue by ryansobol** [textX/textX-LS#36]
My apologies for creating an issue. I'm not sure how best to reach out for support.

I can't seem to get the `textX` VSCode extension to work for my custom language. Here's a screenshot to show how far I've gotten.

<img width="1792" alt="Screen Shot 2020-04-30 at 09 55 13" src="https://user-images.githubusercontent.com/5688/80737895-08dd4580-8ac9-11ea-9593-75ec60603e29.png">

- I've installed the v0.1.2 of the `textX` extension from the marketplace. (latest published)
- My language package depends on v0.1.1 of the `textx-ls-core` package. (latest published)
- As you can see, I can successfully register the language with the `textx` command.
- I've built a wheel and have successfully installed that as a textX language project in VSCode.

What am I missing?

**Comment by danixeee** [comment]
Hello, this is the best and only way for now, no need to apologise! :)

Could you point me to your project (if it is an open-source) and I will try it on my end?

From what I can see, it might be an issue with the file extension that you are using. I haven't tried multiple file extensions (`.tsdoc.py`), so it might be a bug.
Could you try just `.tsdoc` and if that works I will try to find the issue with multiple extensions.

**Comment by ryansobol** [comment]
> Could you point me to your project (if it is an open-source) and I will try it on my end?

That would be super helpful. It's not open source, but I just invited you as a collaborator.

> From what I can see, it might be an issue with the file extension that you are using. I haven't tried multiple file extensions (.tsdoc.py), so it might be a bug. Could you try just .tsdoc and if that works I will try to find the issue with multiple extensions.

Good thought. Unfortunately, that didn't seem to fix the problem. Though it could be one of the problems! 

If possible, I'd like to keep the multiple file extension as the language I'm working on is an embedded comment language for Python. My hope was to leverage VSCode's [grammar injection](https://code.visualstudio.com/api/language-extensions/syntax-highlight-guide#injection-grammars) feature.

**Comment by ryansobol** [comment]
Ping.

**Comment by ipa-mdl** [comment]
TL;DR: Without NodeJS installed (in PATH), the language extension cannot get generated (and installed).
@ryansobol: Do you have NodeJS installed?

The client creates a VSIX file and installs it:
https://github.com/textX/textX-LS/blob/0940a66fb264c7c20f046062606ca6a29af19056/client/src/services/generatorService.ts#L40

In order to create the VSIX file - after some intermediate steps - vsce [gets called](https://github.com/danixeee/textx-gen-vscode/blob/0def8e69a465be331880de054e0d986288705588/textx_gen_vscode/generators.py#L124), whichs needs the NodeJS runtime. However, it is not bundled with VS Code.

@danixeee: Wouldn't it way easier to generate directly into `.vscode/extensions/textx.my_lang-X.Y.Z`?

**Comment by ipa-mdl** [comment]
@danixeee: Or a an (optional) vsix packager that can be used as a drop-in.
The generated VSIX file is just a zip file with two additional manifests..

**Comment by danixeee** [comment]
`vsce` is packaged and shipped within the `textX` extension. 

![image](https://user-images.githubusercontent.com/16227576/90143806-55017c80-dd7e-11ea-97dd-3b329707b653.png)

The problem with `tsdoc` extension was that some characters were not escaped correctly while generating textmate json. It's still an issue, but I didn't have time to resolve it.

On the other hand, I agree that it would be cleaner to have a _vsce_ implementation in python, but I think that it should go to a separate repository.

@ipa-mdl  What are your thoughts?

**Comment by ipa-mdl** [comment]
> vsce is packaged and shipped within the textX extension.

Yes, but it needs NodeJS installed to work. (at least on Windows)
```bat
@IF EXIST "%~dp0\node.exe" (
  "%~dp0\node.exe"  "%~dp0\node_modules\vsce\out\vsce" %*
) ELSE (
  @SETLOCAL
  @SET PATHEXT=%PATHEXT:;.JS;=;%
  node  "%~dp0\node_modules\vsce\out\vsce" %*
)
```

https://github.com/danixeee/textx-gen-vscode/pull/7 is a pure python implementation, which generates the same VSIX content as vsce would (given the templated extension).

I tested it together with #42 on Windows and Linux

**Comment by ryansobol** [comment]
> TL;DR: Without NodeJS installed (in PATH), the language extension cannot get generated (and installed).
> @ryansobol: Do you have NodeJS installed?

Yes

```
~ $ echo $PATH
/usr/local/opt/python@3.8/bin /usr/local/bin /usr/bin /bin /usr/sbin /sbin /Library/Apple/usr/bin
~ $ node --version
v14.7.0
```

**Comment by ipa-mdl** [comment]
> Yes

Ah, so indeed your problem is not related to NodeJs.
Sorry for the noise. It just looked identical.

@danixeee: We should move the discussion to https://github.com/danixeee/textx-gen-vscode/pull/7 then.

> The problem with tsdoc extension was that some characters were not escaped correctly while generating textmate json. It's still an issue, but I didn't have time to resolve it.

Nevertheless, it should be tracked as an issue in https://github.com/danixeee/textx-gen-coloring
